python/ray/experimental/sgd/pytorch/pytorch_trainer.py

```diff
@@ -211,12 +263,30 @@ def restore(self, checkpoint):
         state_id = ray.put(state)
         ray.get([worker.set_state.remote(state_id) for worker in self.workers])
 
-    def shutdown(self):
+    def shutdown(self, force=False):
         """Shuts down workers and releases resources."""
         for worker in self.workers:
-            worker.shutdown.remote()
+            if not force:
+                worker.shutdown.remote()
             worker.__ray_terminate__.remote()
 
+    def _try_resize_workers(self, checkpoint, retries=5):
+        # check available resources
+        self.shutdown(force=True)
+        time.sleep(1)
+        for i in range(retries):
+            resources = ray.available_resources()
+            new_workers = min(resources.get("CPU", 0), self.max_replicas)
+            if self.use_gpu:
+                new_workers = min(resources.get("GPU", 0), new_workers)
+            if new_workers:
+                self._start_workers(int(new_workers))
+                self.restore(checkpoint)
+                return
+            else:
+                time.sleep(2**retries)
```

Comment:
Probably want to print here that you're doing exp backoff and retrying, otherwise could look like a hang